walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,136 @@
+           work on the same event that is shared by multiple attendees) */
+        Map<Event, Boolean> eventsMap = new HashMap<Event, Boolean>();
+        for (Event event : events) { 
+            eventsMap.put(event, true); 
+        }
+
+        Iterator eventsIterator; 
+        for (Map.Entry<String, Boolean> attendeeEntry : attendeesMap.entrySet()) {
+            String attendee = attendeeEntry.getKey();
+
+            eventsIterator = eventsMap.entrySet().iterator();
+            while(eventsIterator.hasNext()) {
+
+                Map.Entry pair = (Map.Entry) eventsIterator.next();
+                if ((Boolean) pair.getValue()) {
+                    Event event = (Event) pair.getKey();
+
+                    Collection<String> eventAttendees = event.getAttendees();
+                    if (eventAttendees.contains(attendee)) {
+                        TimeRange timeRange = event.getWhen();
+                        blockTimeRange(timeRange, request.getDuration(), attendeeEntry.getValue());
+
+                        if (attendeeEntry.getValue()) {
+                            eventsIterator.remove();
+                        }      
+                        // If there are only optional attendees & no times are available for them
+                        if (!attendeeEntry.getValue() && availableTimes.isEmpty()) {
+                            availableTimes.add(TimeRange.WHOLE_DAY);
+                        }
+                    }
+                } 
+            }
+        } 
+        return availableTimes;
+    }
+
+    public void blockTimeRange(TimeRange blockedTimeRange, long meetingDuration, boolean mandatory) {
+        for (int i = 0; i < availableTimes.size(); i++) {
+            TimeRange timeRange = availableTimes.get(i);
+
+            if ((blockedTimeRange.equals(TimeRange.WHOLE_DAY) && !mandatory) 
+                || blockedTimeRange.duration() < meetingDuration) return;
+
+            if (timeRange.overlaps(blockedTimeRange)) {
+
+                TimeRange newTimeRange1 = TimeRange.fromStartEnd(timeRange.start(), 
+                                                blockedTimeRange.start(), false);
+                TimeRange newTimeRange2 = TimeRange.fromStartEnd(blockedTimeRange.end(), 
+                                                timeRange.end(), false);      
+                /*
+                            |--blocked--|
+                        |-----timeRange----|
+                */
+                if (timeRange.contains(blockedTimeRange)) {
+                    System.out.println(timeRange.toString() + " contains " + blockedTimeRange.toString());
+                    if (blockedTimeRange.start() != timeRange.start()) {
+                        createNewTimeRange(i, meetingDuration, newTimeRange1);
+                    }
+                    if (blockedTimeRange.end() != timeRange.end()) {
+                        createNewTimeRange(i + 1, meetingDuration, newTimeRange2);
```

Comment:
Inserting `i + 1` into the same list you're looping through, wouldn't that overwrite the next item that you'll be iterating through?

Reply:
The documentation for ArrayList says that the add() method (which is called in createNewTimeRange()) shifts any subsequent elements to the right. It also shifts elements to the left for the remove() method that's called at the end of my blockTimeRange() function. I figured that this shifting is done before the loop moves on to the next TimeRange in the array,  and also that the size of the array is dynamically updating before the next iteration. When I run the tests it seems like this is in fact the behavior. The only other way I could think to do this is to use an iterator, but I looked online and it seems an iterator isn't the best option to go with

Reply:
I see what you mean, I think that should be ok for your first go.  Just watch out though shifting moves on an array could become expensive.